### PR TITLE
More efficient list indexed attestation endpoint

### PR DESCRIPTION
**Pull request by shayzluf**
[Resolves] #5373
[Resolves] #5486

---

# Description
list indexed attestations endpoint i using beacondb get state function that only returns the lates few epochs states . using the new state-gen will enable to retrieve committees  using old states

- added grouping by blockroot to make the endpoint more efficiant
- added handling for new stategen disabled

### CODE REVIEW

File: beacon-chain/rpc/beacon/attestations.go [hunk 1]
```diff
@@ -91,13 +91,12 @@ func (bs *Server) ListAttestations(
 	}, nil
 }
 
-// ListIndexedAttestations retrieves indexed attestations by target epoch.
-// IndexedAttestationsForEpoch are sorted by data slot by default. Either a target epoch filter
-// or a boolean filter specifying a request for genesis epoch attestations may be used.
+// ListIndexedAttestations retrieves indexed attestations by block root.
+// IndexedAttestationsForEpoch are sorted by data slot by default. Either a start-end epoch
+//filter is used to retrieve blocks with.
```
Comment (farazdagi) [inline beacon-chain/rpc/beacon/attestations.go:96]:
```suggestion
// filter is used to retrieve blocks with.
```

**Comment by prestonvanloon** [comment]
What happens when new state management is not enabled?

**Comment by shayzluf** [comment]
@terencechain i guess it was changed during v0.11 reviews as my old pr got so many conflicts that i just made it again on master without checking that it has beed modified already. anyway this simplifies the code by using get committees from state. changed the description of the pr

**Comment by shayzluf** [comment]
@prestonvanloon added handling for featureconfig

File: beacon-chain/rpc/beacon/attestations.go [hunk 2]
```diff
@@ -6,6 +6,9 @@ import (
 	"strconv"
 	"time"
 
+	stateTrie "github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
+
```
Comment (0xKiwi) [inline beacon-chain/rpc/beacon/attestations.go:11]:
```suggestion
```
You might need to run gaz after this

File: beacon-chain/rpc/beacon/attestations.go [hunk 3]
```diff
@@ -33,6 +36,17 @@ func (s sortableAttestations) Less(i, j int) bool {
 	return s[i].Data.Slot < s[j].Data.Slot
 }
 
+func mapAttestationToBlockRoot(ctx context.Context, atts []*ethpb.Attestation) map[[32]byte][]*ethpb.Attestation {
```
Comment (0xKiwi) [inline beacon-chain/rpc/beacon/attestations.go:39]:
```suggestion
func mapAttestationsByBlockRoot(ctx context.Context, atts []*ethpb.Attestation) map[[32]byte][]*ethpb.Attestation {
```

File: beacon-chain/rpc/beacon/attestations.go [hunk 4]
```diff
@@ -91,13 +105,12 @@ func (bs *Server) ListAttestations(
 	}, nil
 }
 
-// ListIndexedAttestations retrieves indexed attestations by target epoch.
-// IndexedAttestationsForEpoch are sorted by data slot by default. Either a target epoch filter
-// or a boolean filter specifying a request for genesis epoch attestations may be used.
+// ListIndexedAttestations retrieves indexed attestations by block root.
+// IndexedAttestationsForEpoch are sorted by data slot by default. Either a start-end epoch
+// filter is used to retrieve blocks with.
```
Comment (0xKiwi) [inline beacon-chain/rpc/beacon/attestations.go:110]: What should this be? `Either a` but only mentions one option.

File: beacon-chain/rpc/beacon/attestations.go [hunk 5]
```diff
@@ -134,44 +147,46 @@ func (bs *Server) ListIndexedAttestations(
...
-		}
-		activeIndices, err := helpers.ActiveValidatorIndices(attState, epoch)
-		if err != nil {
-			return nil, status.Errorf(
-				codes.Internal,
-				"Could not retrieve active validator indices for epoch %d: %v",
-				epoch,
-				err,
-			)
+	for br, atts := range mappedAttestations {
```
Comment (0xKiwi) [inline beacon-chain/rpc/beacon/attestations.go:154]:
```suggestion
	for blockRoot, atts := range mappedAttestations {
```

File: beacon-chain/rpc/beacon/attestations.go [hunk 6]
```diff
@@ -134,44 +147,46 @@ func (bs *Server) ListIndexedAttestations(
...
+				return nil, status.Errorf(
+					codes.Internal,
+					"Could not retrieve state for attestation data block root %v: %v",
+					br,
+					err,
+				)
+			}
+		} else {
+			attState, err = bs.BeaconDB.State(ctx, br)
+			if err != nil {
```
Comment (0xKiwi) [inline beacon-chain/rpc/beacon/attestations.go:168]: This error is duplicated it seems, maybe you can move it outside of the if statement?

File: beacon-chain/rpc/beacon/attestations.go [hunk 7]
```diff
@@ -134,44 +147,46 @@ func (bs *Server) ListIndexedAttestations(
...
+			committee, err := helpers.BeaconCommitteeFromState(attState, att.Data.Slot, att.Data.CommitteeIndex)
+			if err != nil {
+				return nil, status.Errorf(
+					codes.Internal,
+					"Could not retrieve committees from state %v",
+					err,
+				)
+			}
+			idxAtt := attestationutil.ConvertToIndexed(ctx, att, committee)
+			indexedAtts[i] = idxAtt
```
Comment (0xKiwi) [inline beacon-chain/rpc/beacon/attestations.go:188]: This uses `len(atts)` for `i` but isn't `indexedAtts` `numAttestations` in length?

Comment (shayzluf) [reply beacon-chain/rpc/beacon/attestations.go:188]: thanks fo catching this

**Comment by shayzluf** [comment]
@0xKiwi fixed test to cover the bug you pointed to

File: beacon-chain/rpc/beacon/attestations.go [hunk 8]
```diff
@@ -134,44 +146,40 @@ func (bs *Server) ListIndexedAttestations(
...
-				att.Data.BeaconBlockRoot,
-				err,
-			)
+	attIndex := 0
+	for blockRoot, atts := range mappedAttestations {
+		var attState *stateTrie.BeaconState
+		if !featureconfig.Get().DisableNewStateMgmt {
+			attState, err = bs.StateGen.StateByRoot(ctx, blockRoot)
+		} else {
+			attState, err = bs.BeaconDB.State(ctx, blockRoot)
```
Comment (terencechain) [inline beacon-chain/rpc/beacon/attestations.go:159]: I have a concern about this.  Without new state mgmt, every state between genesis to last finalized epoch gets pruned. What happens if `attState` is nil.

Comment (shayzluf) [reply beacon-chain/rpc/beacon/attestations.go:159]: running it in runtime to test

Comment (shayzluf) [reply beacon-chain/rpc/beacon/attestations.go:159]: it returns an ugly error . but doesnt panic. will return error if stategen disabled

**Comment by shayzluf** [comment]
thanks @terencechain for the feedback

File: beacon-chain/rpc/beacon/attestations.go [hunk 9]
```diff
@@ -117,13 +128,17 @@ func (bs *Server) ListIndexedAttestations(
 	default:
 		return nil, status.Error(codes.InvalidArgument, "Must specify a filter criteria for fetching attestations")
 	}
-	atts := make([]*ethpb.Attestation, 0, params.BeaconConfig().MaxAttestations*uint64(len(blocks)))
+	if featureconfig.Get().DisableNewStateMgmt {
+		return nil, status.Error(codes.Internal, "New state management must be turned on to support historic attestation. Please consider running without --disable-new-state-mgmt flag")
```
Comment (terencechain) [inline beacon-chain/rpc/beacon/attestations.go:132]:
```suggestion
		return nil, status.Error(codes.Internal, "New state management must be turned on to support historic attestation. Please run without --disable-new-state-mgmt flag")
```

**Pull request merged**